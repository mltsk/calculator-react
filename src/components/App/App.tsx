import React, { useEffect, useState } from 'react';

import { KEYS, SPECIAL_BUTTONS, SPECIAL_KEYS } from '../../constants/constants';
import runCalculation from '../../helpers/runCalculation';
import CalculatorContainer from '../CalculatorContainer/CalculatorContainer';
import CalculatorHeader from '../CalculatorHeader/CalculatorHeader';
import Controls from '../Controls/Controls';
import Layout from '../Layout/Layout';

function App() {
  const [calculation, setCalculation] = useState<string | null | undefined>(
    null
  );
  const [result, setResult] = useState<string>('0');

  const handlerClick = (event: React.MouseEvent<HTMLElement>) => {
    if (!(event.target instanceof HTMLElement)) return;
    const char = event.target.textContent;

    if (char === SPECIAL_BUTTONS.RESULT) {
      if (calculation) {
        const result = runCalculation(calculation);

        setResult(result);
      }
    } else if (char === SPECIAL_BUTTONS.DELETE) {
      setCalculation(null);
      setResult('0');
    } else {
      setCalculation((prev) => (prev ? prev + char : char));
    }
  };

  const keyDownHandler = (event: {
    key: string | null | undefined;
    preventDefault: () => void;
  }) => {
    if (!event.key) return;
    if (KEYS.includes(event.key)) {
      setCalculation((prev) => (prev ? prev + event.key : event.key));
    } else if (event.key === SPECIAL_KEYS.ENTER) {
      event.preventDefault();
      if (calculation) {
        const result = runCalculation(calculation);

        setResult(result);
      }
    } else if (event.key === SPECIAL_KEYS.ESCAPE) {
      setCalculation(null);
      setResult('0');
    }
  };

  useEffect(() => {
    window.addEventListener('keydown', keyDownHandler);

    return () => {
      window.removeEventListener('keydown', keyDownHandler);
    };
  });

  return (
    <Layout>
      <CalculatorContainer>
        <CalculatorHeader calculation={calculation} result={result} />
        <Controls handlerClick={handlerClick} />
      </CalculatorContainer>
    </Layout>
  );
}

export default App;
