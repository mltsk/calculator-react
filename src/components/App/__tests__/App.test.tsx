import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import App from '../App';

const calculateExpressionByButtons = (buttons: string): void => {
  buttons.split('').forEach((button) => {
    userEvent.click(screen.getByRole('button', { name: button }));
  });
  userEvent.click(screen.getByRole('button', { name: '=' }));
};

const calculateExpressionByKeys = (keys: string): void => {
  keys.split('').forEach((key) => {
    userEvent.keyboard(`${key}`);
  });
  userEvent.keyboard('{enter}');
};

describe('App component', () => {
  describe('Snapshot', () => {
    it('Should render valid snapshot', () => {
      const { asFragment } = render(<App />);

      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('Main group of tests', () => {
    describe('Calculation using buttons', () => {
      it('Button "C" should clear calculation and result', () => {
        render(<App />);
        calculateExpressionByButtons('6x2');
        expect(screen.getByTestId('calculation')).toHaveTextContent('6x2');
        expect(screen.getByTestId('result')).toHaveTextContent('12');

        userEvent.click(screen.getByRole('button', { name: 'C' }));
        expect(screen.getByTestId('calculation')).toHaveTextContent('');
        expect(screen.getByTestId('result')).toHaveTextContent('0');
      });
      it('Should correctly calculate the expression: 45x10', () => {
        render(<App />);
        calculateExpressionByButtons('45x10');
        expect(screen.getByTestId('result')).toHaveTextContent('450');
      });
      it('Should correctly calculate the expression: 5+10/5', () => {
        render(<App />);
        calculateExpressionByButtons('5+10/5');
        expect(screen.getByTestId('result')).toHaveTextContent('7');
      });
      it('Should correctly calculate the expression: 3x10/5+200%+√36', () => {
        render(<App />);
        calculateExpressionByButtons('3x10/5+200%+√36');
        expect(screen.getByTestId('result')).toHaveTextContent('14');
      });
    });
    describe('Calculation using keys', () => {
      it('Key "Escape" should clear calculation and result', () => {
        render(<App />);
        calculateExpressionByKeys('6x2');
        expect(screen.getByTestId('calculation')).toHaveTextContent('6x2');
        expect(screen.getByTestId('result')).toHaveTextContent('12');

        userEvent.keyboard('{escape}');
        expect(screen.getByTestId('calculation')).toHaveTextContent('');
        expect(screen.getByTestId('result')).toHaveTextContent('0');
      });
      it('Should correctly calculate the expression: 45x10', () => {
        render(<App />);
        calculateExpressionByKeys('45x10');
        expect(screen.getByTestId('result')).toHaveTextContent('450');
      });
      it('Should correctly calculate the expression: 5+10/5', () => {
        render(<App />);
        calculateExpressionByKeys('5+10/5');
        expect(screen.getByTestId('result')).toHaveTextContent('7');
      });
      it('Should correctly calculate the expression: 3x10/5+200%', () => {
        render(<App />);
        calculateExpressionByKeys('3x10/5+200%');
        expect(screen.getByTestId('result')).toHaveTextContent('8');
      });
      it('Should correctly calculate the expression: 4x(300%+7)/4x(20-10)', () => {
        render(<App />);
        calculateExpressionByKeys('4x(300%+7)/4x(20-10)');
        expect(screen.getByTestId('result')).toHaveTextContent('100');
      });
    });
  });
});
