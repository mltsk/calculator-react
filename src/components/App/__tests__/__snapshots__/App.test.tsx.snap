// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Snapshot Should render valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="wrapper"
    >
      <section
        class="calculator"
      >
        <div
          class="header"
        >
          <span
            class="calculation"
            data-testid="calculation"
          />
          <span
            class="result"
            data-testid="result"
          >
            0
          </span>
        </div>
        <div
          class="controls"
        >
          <button
            class="button"
          >
            C
          </button>
          <button
            class="button"
          >
            âˆš
          </button>
          <button
            class="button"
          >
            %
          </button>
          <button
            class="button"
          >
            /
          </button>
          <button
            class="button"
          >
            7
          </button>
          <button
            class="button"
          >
            8
          </button>
          <button
            class="button"
          >
            9
          </button>
          <button
            class="button"
          >
            x
          </button>
          <button
            class="button"
          >
            4
          </button>
          <button
            class="button"
          >
            5
          </button>
          <button
            class="button"
          >
            6
          </button>
          <button
            class="button"
          >
            -
          </button>
          <button
            class="button"
          >
            1
          </button>
          <button
            class="button"
          >
            2
          </button>
          <button
            class="button"
          >
            3
          </button>
          <button
            class="button"
          >
            +
          </button>
          <button
            class="button"
          >
            00
          </button>
          <button
            class="button"
          >
            0
          </button>
          <button
            class="button"
          >
            ,
          </button>
          <button
            class="button result"
          >
            =
          </button>
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
